File: KotlinLab/KotlinLab/src/main/kotlin/ch5/section2_generic/Test2.kt
package org.example.ch5.section2_generic.two

open class Super {
    open fun sayHello(){
        println("Super... sayHello")
    }
}
class Sub: Super(){
    override fun sayHello() {
        println("Sub...")
    }
}

class MyClass<T> {
    fun a(arg: T): T? { return null }
}
class MyClass2<out T>{
//    fun a(arg: T): T? { return null }//error
    fun a(arg: Int): T? { return null }
}
class MyClass3<in T>{
//    fun a(arg: T): T? { return null }//error
    fun a(arg: T): Int? { return null }
}

fun main() {
    //클래스의 상하위이다.. 제네릭 타입이 아니다.. 가능..
    val obj: Super = Sub()
    val obj2: Sub = obj as Sub

    //in, out 이 추가되지 않은 제네릭 타입은 불공변(invariance)
    val obj3 = MyClass<Sub>()
//    val obj4: MyClass<Super> = obj3//error
//    val obj4: MyClass<Sub> = MyClass<Super>()//error

    //covariance
    val obj5: MyClass2<Super> = MyClass2<Sub>()//ok
//    val obj6: MyClass2<Sub> = MyClass2<Super>()//error

    //constravariance
//    val obj7: MyClass3<Super> = MyClass3<Sub>()//error
    val obj8: MyClass3<Sub> = MyClass3<Super>()//ok
}

File: flutter_lab2/lib/ch2/section7/test2.dart
class MyClass {
  int data;
  MyClass(this.data){}//기본(클래스명과 동일한) 생성자..
  MyClass.one(this.data){}//named constructor..
  MyClass.two(this.data){}
}

class MyClass2 {
  int data1 = 10;
  MyClass2(int data1, int data2){
    print("MyClass2 call...");
  }
  //다른 생성자를 호출하는 생성자는 body 를 가질 수 없다..
  // MyClass2.one(int arg): this(arg, 0) { }//error...
  //초기화 목록에는.. 멤버 초기화, 다른 생성자 호출구문이 위치하는데..
  //둘을 같이 사용할 수 없다..
  // MyClass2.one(int arg): this(arg, 0), this.data1 = arg;//error..
  MyClass2.one(int arg): this(arg, 0);
  MyClass2.two(): this.one(0);
}

main(){
  var obj1 = MyClass(10);
  var obj2 = MyClass.one(20);
  var obj3 = MyClass.two(30);

  print("${obj1.data}, ${obj2.data}, ${obj3.data}");//10, 20, 30
}